src/aspire/denoising/denoiser_cov2d.py

```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
         :param covar_opt: The option list for building Cov2D matrix
+        :param overwrite: The option to owerwrite mrcs files for denoised images or not
         """
+        src = self.src
 
-        self.coeffs_noise = self.basis.evaluate_t(self.imgs_noise)
-
-        cov2d = RotCov2D(self.basis)
+        # Initialize the rotationally invariant covariance matrix of 2D images
+        # A fixed batch size is used to go through each image
+        cov2d = BatchedRotCov2D(self.src, self.basis, batch_size=512)
+        if not self.ctf_info:
+            cov2d.ctf_idx = self.ctf_idx
+            cov2d.ctf_fb = self.ctf_fb
 
         default_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
             'iter_callback': [], 'store_iterates': False, 'rel_tolerance': 1e-12,
             'precision': 'float64'}
         covar_opt = fill_struct(covar_opt, default_opt)
-        mean_coeffs_est = cov2d.get_mean(self.coeffs_noise, self.ctf_fb, self.ctf_idx)
+        # Calculate the mean and covariance for the rotationally invariant covariance matrix of 2D images
+        mean_coeffs_est = cov2d.get_mean()
+
+        covar_coeffs_est = cov2d.get_covar(noise_var=self.var_noise, mean_coeff=mean_coeffs_est,
+                                           covar_est_opt=covar_opt)
+        # Denoise the 2D images using the SPCAs from the rotationally invariant covariance matrix
+        # The batch size varies to fit the number of noisy images in each mrcs file
+        _, batch_size = src.group_unique_hist
+
+        img_start = 0
+        for ibatch in range(0, len(batch_size)):
+            imgs_noise = src.images(img_start, batch_size[ibatch])
+            coeffs_noise = self.basis.evaluate_t(imgs_noise.data)
+            logger.info(f'Estimating Cov2D coefficients for image batch {ibatch} of size {batch_size[ibatch]}')
+            coeffs_estim = cov2d.get_cwf_coeffs(coeffs_noise, cov2d.ctf_fb,
+                                                cov2d.ctf_idx[img_start:img_start+batch_size[ibatch]],
+                                                mean_coeff=mean_coeffs_est, covar_coeff=covar_coeffs_est,
+                                                noise_var=self.var_noise)
 
-        covar_coeffs_est = cov2d.get_covar(self.coeffs_noise, self.ctf_fb, self.ctf_idx, mean_coeffs_est,
-                                               noise_var=self.var_noise, covar_est_opt=covar_opt)
+            # Convert Fourier-Bessel coefficients back into 2D images
+            logger.info(f'Converting Cov2D coefficients back to 2D images')
+            imgs_estim = self.basis.evaluate(coeffs_estim)
+            imgs_out = Image(imgs_estim)
+            mrcs_file = os.path.join(self.src.proj_folder, self.starfile_path, self.mrcs_out[img_start])
+            logger.info(f'Saving denoised images to {mrcs_file}')
+            imgs_out.save(mrcs_file, overwrite=overwrite)
+            img_start += batch_size[ibatch]
+
+    def _write_star_file(self):
```

Comment:
Again, we shouldn't have this here. Better to use the save functionality of the `ImageSource` class.

Reply:
Yes. I might change or move to `ImageSource` class.

Reply:
Same as above, a  `DenoisedImageSource` class was added for this purpose in the new commit 23ca25c.